Src/WindowsServer/WindowsServer.Shared/Implementation/DataContracts/AzureInstanceComputeMetadata.cs

```diff
@@ -0,0 +1,177 @@
+                    aimsValue = this.Publisher;
+                    break;
+                case "sku":
+                    aimsValue = this.Sku;
+                    break;
+                case "version":
+                    aimsValue = this.Version;
+                    break;
+                case "vmid":
+                    aimsValue = this.VmId;
+                    break;
+                case "vmsize":
+                    aimsValue = this.VmSize;
+                    break;
+                case "subscriptionid":
+                    aimsValue = this.SubscriptionId;
+                    break;
+                case "resourcegroupname":
+                    aimsValue = this.ResourceGroupName;
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(string.Format(CultureInfo.InvariantCulture, "No field named '{0}' in AzureInstanceComputeMetadata.", fieldName));
+            }
+
+            return aimsValue;
+        }
+
+        /// <summary>
+        /// Because the Azure IMS is on a non-routable IP we need to do some due diligence in our accepting
+        /// values returned from it. This method takes the fieldname and value received for that field, and
+        /// if we can test that value against known limitations of that field we do so here. If the test fails
+        /// we return the empty string, otherwise we return the string given.
+        /// </summary>
+        /// <param name="fieldName">Name of the field acquired from the call to Azure IMS.</param>
+        /// <returns>The value of the field, verified, or the empty string.</returns>
+        internal string VerifyExpectedValue(string fieldName)
+        {
+            string valueToVerify = this.GetValueForField(fieldName);
+            string value = string.Empty;
+            bool valueOk = true;
+
+            if (fieldName.Equals("resourceGroupName", StringComparison.OrdinalIgnoreCase))
+            {
+                valueOk = valueToVerify.Length <= AzureInstanceComputeMetadata.resourceGroupNameLengthMax;
+                valueOk &= valueToVerify.Length >= AzureInstanceComputeMetadata.resourceGroupNameLengthMin;
+                var resGrpMatcher = new Regex(AzureInstanceComputeMetadata.resourceGroupNameValidChars, RegexOptions.None, AzureInstanceComputeMetadata.regexTimeout);
+                valueOk &= resGrpMatcher.IsMatch(valueToVerify);
+                valueOk &= !valueToVerify.EndsWith(".", StringComparison.OrdinalIgnoreCase);
+
+                if (valueOk)
+                {
+                    value = valueToVerify;
+                }
+            }
+            else if (fieldName.Equals("subscriptionId", StringComparison.OrdinalIgnoreCase))
+            {
+                Guid g = new Guid();
+                valueOk = Guid.TryParse(valueToVerify, out g);
+
+                if (valueOk)
+                {
+                    value = valueToVerify;
+                }
```

Comment:
what happens when valueOk is false?

Reply:
If you look further down at line ~170 you will see that the valueOk flag is tested for false, and if it is false a log message is written to the WindowsServerEventSource object. Also, if the valueOk flag is false, the valueToVerify is never returned, effectively ‘sanitizing’ the field.